### PR TITLE
gdlauncher-carbon: init at 2.0.20

### PR DESCRIPTION
**Pull request by huantianad**
## Description of changes
Package request issue: #269067

Packaged by extracting `app.asar` as well as binary blob from provided AppImage, and wrapping with our own electron.

Haven't done much testing, so putting this as a draft for now, just to get some eyes on this. I also haven't added any of the libraries that MC itself needs. I wonder if we could borrow some code from `prismlauncher`.

If wanted, we could also provide a derivation built from source; however as mentioned in the request issue thread, it won't have all the features as the pre-built version comes with an API key for their own servers.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by huantianad** [comment]
@PassiveLemon You might be interested!

File: pkgs/by-name/gd/gdlauncher-carbon/package.nix [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+  '';
+
+  postConfigure =
+    let
+      libPath = lib.makeLibraryPath [
+        libxcb
+      ];
+      binPath = lib.makeBinPath [
+        xdg-utils
+      ];
```
Comment (huantianad) [inline pkgs/by-name/gd/gdlauncher-carbon/package.nix:54]: I'm not 100% sure this is needed, but I'm assuming that app does depend on xdg-utils to open up browser pages, including for signin. Would have to test on a system without it though.

**Comment by PassiveLemon** [comment]
I'll check it out later, looks promising though

**Comment by PassiveLemon** [comment]
Upon running GDL, it returns: `cannot execute /nix/store/lcvhr3s23c2i53ackxh8c39lggcfs0mq-gdlauncher-carbon-2.0.1/share/gdlauncher-carbon/resources/binaries/core_module: You are trying to run an unpatched binary on nixos, but you have not configured NIX_LD or NIX_LD_x86_64-linux. See https://github.com/Mic92/nix-ld for more details`
The icon is also not used.
Seems like something is up. My personal package works fine however.

**Comment by huantianad** [comment]
Wups I've been testing this with mix-ld 🤦

**Comment by PassiveLemon** [comment]
> Also wdym about the icon?

The icon is simply not in my app launcher.
image

**Comment by huantianad** [comment]
Man I thought I fixed that with my recent force pushes, can you check to see if the icons were installed in the right folders?

**Comment by PassiveLemon** [comment]
I see icons in both `$out/share/icons/hicolor/128x128/apps/gdlauncher-carbon.png` and 512x512 and I can open both in feh so I'm not really sure what the issue is there.

In my package, the icon was installed in `$out/share/pixmaps/gdlauncher-carbon.png` so I added a line to install in there in the derivation and it does find the icon after that.

**Comment by huantianad** [comment]
That's strange, pixmap shouldn't be the place to put these icons tho. Maybe I should also add a 48x48 and 96x96 size?

**Comment by PassiveLemon** [comment]
Could be, the size, yeah. It could also just be my app launcher so I'll add some smaller sizes and test those

**Comment by huantianad** [comment]
I've fixed the core binary issue as well as added some of the libraries from `prismlauncher`. Not sure if all of them are necessary, or if the ones I skipped are also needed. Perhaps @Scrumplex can give some insight here.

**Comment by Scrumplex** [comment]
`glfw` and `openal` might not be needed here, as Prism Launcher has a feature to override the versions shipped by LWJGL, while I assume GDLauncher does not

**Comment by PassiveLemon** [comment]
Tested the package, it works

**Comment by huantianad** [comment]
> `glfw` and `openal` might not be needed here, as Prism Launcher has a feature to override the versions shipped by LWJGL, while I assume GDLauncher does not

Makes sense, I've mentioned the feature to the dev and they seem interested in possibly adding in the future. For now, removing.

File: pkgs/by-name/gd/gdlauncher-carbon/package.nix [hunk 2]
```diff
@@ -0,0 +1,119 @@
+{ lib
+, stdenv
```
Comment (FliegendeWurst) [inline pkgs/by-name/gd/gdlauncher-carbon/package.nix:2]: Please format with the new formatter.

File: pkgs/by-name/gd/gdlauncher-carbon/package.nix [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      desktopName = "GDLauncher";
+      comment = finalAttrs.meta.description;
+      categories = [ "Game" ];
+      keywords = [ "launcher" "mod manager" "minecraft" ];
+      mimeTypes = [ "x-scheme-handler/gdlauncher" ];
+    })
+  ];
+
+  meta = {
+    description = "A simple, yet powerful Minecraft custom launcher with a strong focus on the user experience";
```
Comment (FliegendeWurst) [inline pkgs/by-name/gd/gdlauncher-carbon/package.nix:112]:
```suggestion
    description = "Simple, yet powerful Minecraft custom launcher with a strong focus on the user experience";
```

File: pkgs/by-name/gd/gdlauncher-carbon/package.nix [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    xorg.libxcb
+  ];
+
+  installPhase = ''
+    runHook preInstall
+
+    mkdir -p $out/share/gdlauncher-carbon/resources
+    cp -r $src/resources/{binaries,app.asar} $out/share/gdlauncher-carbon/resources/
+
+    # The provided icon is a bit large for some systems, so make smaller ones
```
Comment (FliegendeWurst) [inline pkgs/by-name/gd/gdlauncher-carbon/package.nix:52]: I think this is done automatically by all relevant icon users?

**Review by FliegendeWurst — DISMISSED** [review]

File: pkgs/by-name/gd/gdlauncher-carbon/package.nix (hunk 4 above)
Comment (huantianad) [reply pkgs/by-name/gd/gdlauncher-carbon/package.nix:58]: Perhaps I can try again; last time I checked, it didn't seem to work on KDE Plasma.

File: pkgs/by-name/gd/gdlauncher-carbon/package.nix [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+  makeWrapper,
+  copyDesktopItems,
+  autoPatchelfHook,
+  xorg,
+  libpulseaudio,
+  libGL,
+  udev,
+  xdg-utils,
+  electron,
+  addOpenGLRunpath,
```
Comment (FliegendeWurst) [inline pkgs/by-name/gd/gdlauncher-carbon/package.nix:16]: Use addDriverRunPath instead, see #275241

**Comment by Pandapip1** [comment]
Currently failing with:

```
       error: evaluation aborted with the following error message: 'lib.customisation.callPackageWith: Function called without required argument "addOpenGLRunpath" at /ofborg/checkout/1/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/ofborg-evaluator-3/pkgs/by-name/gd/gdlauncher-carbon/package.nix:16'
```

**Comment by huantianad** [comment]
By the way, if someone could help co-maintain this package, would be appreciated, as I don't use it myself at all.

**Comment by TsubakiDev** [comment]
I have a lot of free time, can I apply to be a co-maintainer of this package?

**Comment by FliegendeWurst** [comment]
> I have a lot of free time, can I apply to be a co-maintainer of this package?

Yes, certainly :)

Since you are not listed in maintainer-list.nix yet, huantianad can add your details there and then add you to meta.maintainers of this package.

**Comment by TsubakiDev** [comment]
Thanks (⁠｡⁠･⁠ω⁠･⁠｡⁠)⁠ﾉ⁠♡
I'm happy to be the maintainer of this package (⁠◍⁠•⁠ᴗ⁠•⁠◍⁠)⁠❤
________________________________
发件人: Arne Keller ***@***.***>
发送时间: 2024年12月8日 20:32
收件人: NixOS/nixpkgs ***@***.***>
抄送: Daniel Wang ***@***.***>; Comment ***@***.***>
主题: Re: [NixOS/nixpkgs] gdlauncher-carbon: init at 2.0.20 (PR #297096)


I have a lot of free time, can I apply to be a co-maintainer of this package?

Yes, certainly :)

Since you are not listed in maintainer-list.nix yet, huantianad can add your details there and then add you to meta.maintainers of this package.

**Comment by huantianad** [comment]
Not sure what the procedure for this is, @TsubakiDev can you let me know what you want your entry in https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix to look like? Comment at the start has instructions.

I'll make a commit with that myself, or I suppose if you'd like you could make the commit in your own fork and I'll cherry-pick it into this PR branch (though that's probably unnecessarily complex).

**Comment by TsubakiDev** [comment]
~~~text
TsubakiDev = {
    email = ***@***.***";
    github = "TsubakiDev";
    githubId = 132794625;
    name = "Daniel Wang";
  };
~~~

This is my entry at `maintainer-list.nix`.

于 2024年12月9日 GMT+08:00 14:12:35，huantian ***@***.***> 写道：

**Comment by TsubakiDev** [comment]
~~~
TsubakiDev = {
    email = ***@***.***";
    github = "TsubakiDev";
    githubId = 132794625;
    name = "Daniel Wang";
  };
~~~

This is my entry at `maintainer-list.nix`.

**Comment by TsubakiDev** [comment]
It seems that the email cannot display markdown and my email address.

My email address is i at tsubakidev dot cc

**Comment by TsubakiDev** [comment]
> Not sure what the procedure for this is, @TsubakiDev can you let me know what you want your entry in https://github.com/NixOS/nixpkgs/blob/master/maintainers/maintainer-list.nix to look like? Comment at the start has instructions.
> 
> I'll make a commit with that myself, or I suppose if you'd like you could make the commit in your own fork and I'll cherry-pick it into this PR branch (though that's probably unnecessarily complex).

```
TsubakiDev = {
    email = i@tsubakidev.cc";
    github = "TsubakiDev";
    githubId = 132794625;
    name = "Daniel Wang";
  };
```
This is my entry at `maintainer-list.nix`.

**Review by TsubakiDev — APPROVED** [review]

**Review by TsubakiDev — APPROVED** [review]

File: pkgs/by-name/gd/gdlauncher-carbon/package.nix [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+  nativeBuildInputs = [
+    graphicsmagick
+    makeWrapper
+    copyDesktopItems
+    autoPatchelfHook
+  ];
+
+  buildInputs = [
+    xorg.libxcb
+  ];
```
Comment (FliegendeWurst) [inline pkgs/by-name/gd/gdlauncher-carbon/package.nix:50]:
```suggestion
  buildInputs = [
    xorg.libxcb
    stdenv.cc.cc.lib
  ];

  strictDeps = true;
```
One of the binaries requires libgcc_s. Enable `strictDeps` to make sure we are explicit about dependencies.

**Review by TsubakiDev — APPROVED** [review]

**Pull request merged**